Fix logic for terminating a child process spamming us with IPC

cdumez opened the pull request:
#### ad1a65843e9e1fd4264ac34a4777729fa43f943f
<pre>
Fix logic for terminating a child process spamming us with IPC
<a href="https://bugs.webkit.org/show_bug.cgi?id=277746">https://bugs.webkit.org/show_bug.cgi?id=277746</a>

Reviewed by Per Arne Vollan.

Make the following fixes to the logic:
1. Once the message queue size reaches maxPendingIncomingMessagesKillingThreshold, only
   call dispatchToClientWithIncomingMessagesLock() once to terminate the child process
   on the main thread. We used to dispatch a task for each message over the threshold,
   which was unnecessary.
2. After calling m_client-&gt;requestRemoteProcessTermination(), check if Connection::m_wasKilled
   is true. If it&apos;s not, call Connection::kill(). This used to be guaranteed when using XPC
   services since calling requestRemoteProcessTermination() would terminate the xpc connection.
   However, when using ExtensionKit (on iOS currently), it no longer kills the XPC connection
   since it is not required to terminate the child extension. This is useful because having
   Connection::m_wasKilled set to true causes Connection::enqueueIncomingMessage() do early
   return instead of doing work for every new incoming message while waiting for the child
   process to exit.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::enqueueIncomingMessage):

Canonical link: <a href="https://commits.webkit.org/281969@main">https://commits.webkit.org/281969@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0cc0f70080ac9f84a9bf66485f4fd22f3dab7b9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61567 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40917 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14150 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65515 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12108 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48601 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12381 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49680 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8414 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64635 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38020 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53326 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30518 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34682 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11017 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56489 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10857 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67241 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5506 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57065 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5532 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53281 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57283 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4533 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9274 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36719 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37804 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38898 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37549 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/Platform/IPC/Connection.cpp:
```diff
@@ -1335,14 +1335,18 @@ void Connection::enqueueIncomingMessage(UniqueRef<Decoder> incomingMessage)
+                        return;
+                    protectedThis->m_client->requestRemoteProcessTermination();
+                    if (!protectedThis->m_wasKilled)
+                        protectedThis->kill();
```
Is this method available when `USE(EXTENSIONKIT_PROCESS_TERMINATION)`?

cdumez replied:
Yes, good point. I'll re-work the patch.

pvollan reviewed (approved):
Nice! R=me.

cdumez committed:
Fix logic for terminating a child process spamming us with IPC
https://bugs.webkit.org/show_bug.cgi?id=277746

Reviewed by Per Arne Vollan.

Make the following fixes to the logic:
1. Once the message queue size reaches maxPendingIncomingMessagesKillingThreshold, only
   call dispatchToClientWithIncomingMessagesLock() once to terminate the child process
   on the main thread. We used to dispatch a task for each message over the threshold,
   which was unnecessary.
2. After calling m_client->requestRemoteProcessTermination(), check if Connection::m_wasKilled
   is true. If it's not, call Connection::kill(). This used to be guaranteed when using XPC
   services since calling requestRemoteProcessTermination() would terminate the xpc connection.
   However, when using ExtensionKit (on iOS currently), it no longer kills the XPC connection
   since it is not required to terminate the child extension. This is useful because having
   Connection::m_wasKilled set to true causes Connection::enqueueIncomingMessage() do early
   return instead of doing work for every new incoming message while waiting for the child
   process to exit.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::enqueueIncomingMessage):

Canonical link: https://commits.webkit.org/281969@main
